### ISSUE TITLE
AttributeError: 'str' object has no attribute 'append'

### ISSUE THREAD

**Issue by dxplay120** [facebook/redex#529]
python2 redex/tools/hprof/dump_classes_from_hprof.py --hprof ~/SOMEDUMP.hprof > ~/list_of_classes.txt

Traceback (most recent call last):
  File "redex/tools/hprof/dump_classes_from_hprof.py", line 1437, in <module>
    hp = parse_filename(args.hprof)
  File "redex/tools/hprof/dump_classes_from_hprof.py", line 63, in parse_filename
    return parse_hprof_dump(open(filename, "rb"))
  File "redex/tools/hprof/dump_classes_from_hprof.py", line 37, in parse_hprof_dump
    tag.append(byte)
AttributeError: 'str' object has no attribute 'append'

**Comment by amnn** [comment]
Hi @dxplay120, thanks for spotting this issue! The fix is landing at the moment.

**Comment by amnn** [comment]
Should be fixed by #530 

I've done some local testing, but do not have an hprof handy to test end-to-end.  Please could you try it out and see if it works?

**Comment by dxplay120** [comment]
> Should be fixed by #530
> 
> I've done some local testing, but do not have an hprof handy to test end-to-end. Please could you try it out and see if it works?

@amnn thanks!.but got a new issue

  File "redex/tools/hprof/dump_classes_from_hprof.py", line 1437, in <module>
    hp = parse_filename(args.hprof)
  File "redex/tools/hprof/dump_classes_from_hprof.py", line 63, in parse_filename
    return parse_hprof_dump(open(filename, "rb"))
  File "redex/tools/hprof/dump_classes_from_hprof.py", line 49, in parse_hprof_dump
    record = Record.read_from_stream(hprof_data, instream)
  File "redex/tools/hprof/dump_classes_from_hprof.py", line 182, in read_from_stream
    if tag == HprofTag.STRING.value:
AttributeError: 'int' object has no attribute 'value'

**Comment by amnn** [comment]
@dxplay120, this tool is intended to be run with Python 3, I notice in your original message that you were running it with Python 2.  `HprofTag` is an enum, which is a [Python 3.4+ feature](https://docs.python.org/3/library/enum.html).  Could you try rerunning with a recent version of Python 3?

I have checked with a small program:

```python
import enum

class Foo(enum.Enum):
  BAR = 0x1

exit(Foo.BAR.value)
```

This works with Python 3 but complains about not finding the `enum` module in Python 2.

**Comment by dxplay120** [comment]
> @dxplay120, this tool is intended to be run with Python 3, I notice in your original message that you were running it with Python 2. `HprofTag` is an enum, which is a [Python 3.4+ feature](https://docs.python.org/3/library/enum.html). Could you try rerunning with a recent version of Python 3?
> 
> I have checked with a small program:
> 
> ```python
> import enum
> 
> class Foo(enum.Enum):
>   BAR = 0x1
> 
> exit(Foo.BAR.value)
> ```
> 
> This works with Python 3 but complains about not finding the `enum` module in Python 2.

When I use python3 it will also report an error

```
  File "redex/tools/hprof/dump_classes_from_hprof.py", line 1437, in <module>
    hp = parse_filename(args.hprof)
  File "redex/tools/hprof/dump_classes_from_hprof.py", line 63, in parse_filename
    return parse_hprof_dump(open(filename, "rb"))
  File "redex/tools/hprof/dump_classes_from_hprof.py", line 49, in parse_hprof_dump
    record = Record.read_from_stream(hprof_data, instream)
  File "redex/tools/hprof/dump_classes_from_hprof.py", line 180, in read_from_stream
    data = array(b"B")
TypeError: array() argument 1 must be a unicode character, not bytes
```

**Comment by dxplay120** [comment]
when i change data = array(b"B") to data = array("B") or data = array(u"B") it also report an error
`redex/tools/hprof/dump_classes_from_hprof.py:49: DeprecationWarning: fromstring() is deprecated. Use frombytes() instead.
  record = Record.read_from_stream(hprof_data, instream)
redex/tools/hprof/dump_classes_from_hprof.py:867: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.
  string_record = StringRecord.create(tag, time_offset_us, data)
Traceback (most recent call last):
  File "redex/tools/hprof/dump_classes_from_hprof.py", line 1437, in <module>
    hp = parse_filename(args.hprof)
  File "redex/tools/hprof/dump_classes_from_hprof.py", line 63, in parse_filename
    return parse_hprof_dump(open(filename, "rb"))
  File "redex/tools/hprof/dump_classes_from_hprof.py", line 53, in parse_hprof_dump
    hprof_data.resolve()
  File "redex/tools/hprof/dump_classes_from_hprof.py", line 970, in resolve
    clazz.resolve(self)
  File "redex/tools/hprof/dump_classes_from_hprof.py", line 622, in resolve
    setattr(self.fields, name, static_field.value)
TypeError: attribute name must be string, not 'bytes'`

**Comment by amnn** [comment]
I can repro the first issue locally as well.  Let me get an hprof to run through the script so I can get it to a fully working state.  I can then get back to you!

**Comment by amnn** [comment]
I now have it working locally, will submit it as another PR.

**Comment by amnn** [comment]
@dxplay120, could you take a look at the PR I just linked (#531)?  It'll take a little longer to get landed as I want to get it reviewed by someone else (this is not an area of the codebase I'm very familiar with), but I was able to run the script all the way through on an hprof file, locally using Python 3, so it should hopefully unblock you.

**Comment by amnn** [comment]
@dxplay120, the fix is now landed.  I also checked with the team and it looks like this tool doesn't see much active use internally, so support may be spotty (there are some issues with hprofs from later Android versions, for example) -- just a warning, if you are planning on relying on it heavily.

**Comment by dxplay120** [comment]
> @dxplay120, the fix is now landed. I also checked with the team and it looks like this tool doesn't see much active use internally, so support may be spotty (there are some issues with hprofs from later Android versions, for example) -- just a warning, if you are planning on relying on it heavily.

@amnn Thanks!  I tried with the fixed code, and there was a new problem

```
python3 tools/hprof/dump_classes_from_hprof.py --hprof ~/SOMEDUMP.hprof > ~/list_of_classes.txt
Traceback (most recent call last):
  File "tools/hprof/dump_classes_from_hprof.py", line 1437, in <module>
    hp = parse_filename(args.hprof)
  File "tools/hprof/dump_classes_from_hprof.py", line 63, in parse_filename
    return parse_hprof_dump(open(filename, "rb"))
  File "tools/hprof/dump_classes_from_hprof.py", line 53, in parse_hprof_dump
    hprof_data.resolve()
  File "tools/hprof/dump_classes_from_hprof.py", line 970, in resolve
    clazz.resolve(self)
  File "tools/hprof/dump_classes_from_hprof.py", line 613, in resolve
    static_field.value = hprof_data.object_id_dict[static_field.value]
KeyError: 7274596
```

SOMEDUMP.hprof file comes from Android 5.1 device

**Comment by amnn** [comment]
@dxplay120, I've noticed that @agampe is working on a mitigation for this issue, so may be able to speak with more authority on what to expect from this script, but this does seem like the kind of problem that I was alluding to before:

> support may be spotty (there are some issues with hprofs from later Android versions, for example)

For the purposes of this statement, Android 5.1 may be later than this tool supports.

**Comment by agampe** [comment]
Other issues on Github describe this as "Only Dalvik (Android 4 and earlier) is officially supported." It seems later HPROF files generated by ART may not be technically correct/complete.

Use the master version at or after 1bee36fa and `--allow_missing_ids`, in case this is just one or two missing objects.
